File: backend/routes/report_routes.py
"""Report routes — end-user submit + admin queue."""
from __future__ import annotations
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel, Field
from typing import Literal

from middleware.auth_middleware import require_auth
from services import report_service
from utils.rate_limit import check_and_record

router = APIRouter(prefix="/v1", tags=["reports"])


ReportTargetType = Literal["listing", "user", "review", "message"]
ReportReason = Literal["spam", "offensive", "scam", "wrong_category", "other"]


class ReportBody(BaseModel):
    target_type: ReportTargetType
    target_id: str
    reason: ReportReason
    description: str | None = Field(None, max_length=1000)


class ResolveBody(BaseModel):
    action: Literal["takedown", "warn", "ban", "none"] = "none"
    note: str | None = None


class DismissBody(BaseModel):
    reason: str | None = None


def _require_admin(u):
    if "admin" not in (u.roles or []):
        raise HTTPException(403, "Admin only")


@router.post("/reports")
async def create_report(body: ReportBody, request: Request, user=Depends(require_auth)):
    # Rate limit 10 / hour / user
    allowed, reset_in = check_and_record(f"report:{user.id}", limit=10, window_seconds=3600)
    if not allowed:
        raise HTTPException(429, f"Too many reports. Try again in {reset_in} seconds.")
    return await report_service.create_report(user.id, body.model_dump(exclude_none=True))


@router.get("/admin/reports")
async def admin_list_reports(
    status: str | None = None,
    target_type: str | None = None,
    cursor: str | None = None,
    limit: int = Query(30, ge=1, le=100),
    user=Depends(require_auth),
):
    _require_admin(user)
    return await report_service.list_reports(status, target_type, cursor, limit)


@router.post("/admin/reports/{rid}/resolve")
async def admin_resolve(rid: str, body: ResolveBody, user=Depends(require_auth)):
    _require_admin(user)
    return await report_service.resolve_report(rid, user.id, body.action, body.note)


@router.post("/admin/reports/{rid}/dismiss")
async def admin_dismiss(rid: str, body: DismissBody | None = None, user=Depends(require_auth)):
    _require_admin(user)
    reason = body.reason if body else None
    return await report_service.dismiss_report(rid, user.id, reason)

File: backend/src/routes/report.routes.js
const express = require('express');
const { requireAuth } = require('../middleware/auth.middleware');
const reportService = require('../services/report.service');
const { checkAndRecord } = require('../utils/rateLimit');
const { catchAsync } = require('../utils/helpers');
const ApiError = require('../utils/ApiError');

const router = express.Router();

const requireAdmin = (req, res, next) => {
  const roles = req.user.roles || [];
  if (!roles.includes('admin')) {
    return next(ApiError.forbidden('Admin only'));
  }
  next();
};

router.post('/reports', requireAuth, catchAsync(async (req, res) => {
  const { target_type, target_id, reason, description } = req.body;
  if (!target_type || !target_id || !reason) {
    throw ApiError.badRequest('target_type, target_id and reason are required');
  }
  if (!['listing', 'user', 'review', 'message'].includes(target_type)) {
    throw ApiError.badRequest('Invalid target_type');
  }
  if (!['spam', 'offensive', 'scam', 'wrong_category', 'other'].includes(reason)) {
    throw ApiError.badRequest('Invalid reason');
  }

  // Rate limit: 10 reports per hour per user
  const userId = req.user._id.toString();
  const { allowed, remaining } = checkAndRecord(`report:${userId}`, 10, 3600);
  if (!allowed) {
    throw new ApiError(429, `Too many reports. Try again in ${remaining} seconds.`);
  }

  const result = await reportService.createReport(userId, {
    target_type,
    target_id,
    reason,
    description: description ? String(description).slice(0, 1000) : null,
  });

  res.json(result);
}));

router.get('/admin/reports', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const { status, target_type, cursor } = req.query;
  const limit = Math.max(1, Math.min(100, parseInt(req.query.limit || 30, 10)));

  const result = await reportService.listReports(status || null, target_type || null, cursor || null, limit);
  res.json(result);
}));

router.post('/admin/reports/:rid/resolve', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const { action = 'none', note } = req.body;
  if (!['takedown', 'warn', 'ban', 'none'].includes(action)) {
    throw ApiError.badRequest('Invalid action');
  }

  const result = await reportService.resolveReport(req.params.rid, req.user._id.toString(), action, note || null);
  res.json(result);
}));

router.post('/admin/reports/:rid/dismiss', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const reason = req.body?.reason || null;
  const result = await reportService.dismissReport(req.params.rid, req.user._id.toString(), reason);
  res.json(result);
}));

module.exports = router;
